collectors/python.d.plugin/freeswitch/freeswitch.chart.py

```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+# Description: nginx netdata python.d module
+# Author: Koldo Aingeru Marcos (paulfantom)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+
+try:
+    import ESL
+    HAS_ESL = True
+except ImportError:
+    HAS_ESL = False
+
+import json
+
+from bases.FrameworkServices.SimpleService import SimpleService
+
+
+DEFAULT_HOST = '127.0.0.1'
+DEFAULT_PORT = 8021
+DEFAULT_PASSWORD = 'ClueCon'
+
+ORDER = [
+    'calls',
+    'registrations',
+    'gateways',
+    'profiles'
+]
+
+
+CHARTS = {
+    'calls': {
+        'options': [None, 'Active Calls', 'calls', 'calls', 'freeswitch.calls', 'line'],
+        'lines': [
+            ['calls','calls','absolute']
+        ]
+    },
+    'registrations': {
+        'options': [None, 'Registrations', 'registrations', 'registrations', 'freeswitch.registrations', 'line'],
+        'lines': [
+            ['registrations','registrations','absolute']
+        ]
+    },
+    'gateways': {
+        'options': [None, 'Gateways', 'gateways', 'gateways', 'freeswitch.gateways', 'line'],
+        'lines': [
+            ['UNREGED','UNREGED','absolute'],
+            ['TRYING','TRYING','absolute'],
+            ['REGISTER','REGISTER','absolute'],
+            ['REGED','REGED','absolute'],
+            ['UNREGISTER','UNREGISTER','absolute'],
+            ['FAILED','FAILED','absolute'],
+            ['FAIL_WAIT','FAIL_WAIT','absolute'],
+            ['EXPIRED','EXPIRED','absolute'],
+            ['NOREG','NOREG','absolute'],
+            ['TIMEOUT','TIMEOUT','absolute'],
+        ]
+    },
+    'profiles': {
+        'options': [None, 'Active Profiles', 'profiles', 'profiles', 'freeswitch.profiles', 'line'],
+        'lines': [
+            ['RUNNING','RUNNING','absolute'],
+            ['DOWN','DOWN','absolute'],
+        ]
+    }
+}
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+
+        self.order = ORDER
+        self.definitions = CHARTS
+
+        self.host = configuration.get('host', DEFAULT_HOST)
+        self.port = configuration.get('port', DEFAULT_PORT)
+        self.password = configuration.get('password', DEFAULT_PASSWORD)
+
+
+    def check(self):
+        if not HAS_ESL:
+            self.error("'ESL' package is needed to use freeswitch module, you can install with with pip install python-ESL")
+            return False
+
+        try:
+            self.fs = ESL.ESLconnection(self.host, self.port, self.password)
+            if not self.fs.connected():
+                raise ValueError('Not connected')
+        except ValueError:
```

Comment:
How about to catch the error and log it?

```python
try:
    self.fs = ESL.ESLconnection(self.host, self.port, self.password)
except ValueError as error:
    self.error('SOME TEXT WITH error')
    return False
```

Are you sure about ValueError btw? I think we need some error from ESL package

Reply:
Based on the (very limited) ESL documentation I've found, it looks like the `ESL.ESLconnection` class won't throw an error during initialization, but may not be connected once initialized.

So this try/except block appears to just be flow control, in which case it should instead be rewritten as:

```python
self.fs = ESL.ESLconnection(self.host, self.port, self.password)

if not self.fs.connected():
    self.error("not connected")
    return false
```

Or, alternatively if I'm wrong about the `ESL.ESLconnection` constructor not being able to throw an error:

```python
try:
   self.fs = ESL.ESLconnection(self.host, self.port, self.password)
except Exception as error:
    self.error("SOME TEXT WITH error")
    return false

if not self.fs.connected():
    self.error("not connected")
    return false
```

With `Exception` replaced by whatever base class ESL uses for errors.

Reply:
> it looks like the ESL.ESLconnection class won't throw an error during initialization

@brablc please test it

Reply:
Seems that it won't throw any exception or error, I changed it to the first approach and seems to work

```
2019-10-09 08:22:53: python.d DEBUG: plugin: main: job initialization: 'freeswitch localhost' => ['OK']
2019-10-09 08:22:53: python.d DEBUG: plugin: main: module status: 'freeswitch' => [OK] (jobs: 1)
2019-10-09 08:22:53: python.d ERROR: freeswitch: localhost: Cannot connect to FreeSWITCH with provided data, check that the service is running
2019-10-09 08:22:53: python.d INFO: freeswitch: localhost: check() => [FAILED]
2019-10-09 08:22:53: python.d INFO: plugin: main: FINISHED
```

Reply:
but

> conn.connected()

return False and again no exception?

Reply:
That's it...

```
>>> import ESL
>>> conn = ESL.ESLconnection("127.0.0.1", 4321, "swordfish")
>>> if not conn.connected():
...     print 'Not connected'
... 
Not connected
>>>
```

Reply:
I'm using another library for another FreeSWITCH related project, (greenswitch)[https://github.com/EvoluxBR/greenswitch] which raises exceptions:

```
>>> import greenswitch
>>> fs = greenswitch.InboundESL(host='127.0.0.1', port=1234, password='swordfish')
>>> fs.connect()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/greenswitch/esl.py", line 236, in connect
    self.sock.connect((self.host, self.port))
  File "/usr/lib/python2.7/dist-packages/gevent/socket.py", line 342, in connect
    raise error(err, strerror(err))
socket.error: [Errno 111] Connection refused
>>> 
```

I can rewrite the module using it, I think that I started with `python-ESL` because it was what the documentation of FreeSWITCH uses and maybed it has lees dependencies but might be worth switch to it.

Reply:
> Now that I think about it, with this library it could get more things

I m not clear, which library is better? ESL or greenswitch?

Reply:
Just my two cents, but unless there's some major issue, it sounds like Greenswitch is probably a better option here.  Less maintained isn't great, but pure Python is (it makes it less likely that people won't be able to install it than a C extension), the extra data would be nice, and the fact that it throws proper exceptions should make the handling of the data collection more robust and less confusing.